ion-hash/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+
+//! Implement Ion Hash as per the [spec](https://amzn.github.io/ion-hash/docs/spec.html)
+//!
+//! ## Examples
+//! ```rust
+//! use ion_rs::value::loader::{self, Loader};
+//! use ion_rs::result::IonResult;
+//!
+//! # fn main() -> IonResult<()> {
+//!   let loader = loader::loader();
+//!   let elem = loader.iterate_over(b"\"hello world\"")?.next().unwrap()?;
+//!   let digest = ion_hash::sha256(&elem);
+//!   println!("{:?}", digest);
+//! # Ok(())
+//! # }
+//! ```
+
+use ion_rs::{
+    value::{owned::OwnedElement, Element},
+    IonType,
+};
+
+use sha2::{digest::Output, Digest, Sha256};
+
+pub fn sha256(elem: &OwnedElement) -> Vec<u8> {
+    let mut hasher = IonHasher::new(Sha256::new());
+    let result = hasher.visit_element(elem);
+    Vec::from(result.as_slice())
+}
+
+/// Bytes markers as per the spec.
+mod markers {
+    /// single byte begin marker
+    pub(crate) const B: u8 = 0x0B;
+    /// single byte end marker
+    pub(crate) const E: u8 = 0x0E;
+    /// single byte escape
+    pub(crate) const ESC: u8 = 0x0C;
+    /// type qualifier octet consisting of a four-bit type code T followed by a four-bit qualifier Q
+    /// (this varies per type)
+    pub(crate) const TQ_STRING: u8 = 0x80;
+}
+
+// TODO: In the fullness of time, we will have a IonHashReader and a
+// IonHashWriter. This will allow reading/writing a value *while* computing a
+// hash. For now, we provide a standalone hasher using the OwnedElement API.
+pub struct IonHasher<D>
+where
+    D: Digest,
+{
+    hasher: D,
+}
+
+impl<D> IonHasher<D>
+where
+    D: Digest,
+{
+    fn new(hasher: D) -> IonHasher<D> {
+        IonHasher { hasher }
+    }
+
+    fn visit_element(&mut self, elem: &OwnedElement) -> Output<D> {
+        let serialized_bytes = match elem.ion_type() {
+            IonType::Null => todo!(),
+            IonType::Boolean => todo!(),
+            IonType::Integer => todo!(),
+            IonType::Float => todo!(),
+            IonType::Decimal => todo!(),
+            IonType::Timestamp => todo!(),
+            IonType::Symbol => todo!(),
+            IonType::String => self.visit_string(elem.as_str().unwrap()),
+            IonType::Clob => todo!(),
+            IonType::Blob => todo!(),
+            IonType::List => todo!(),
+            IonType::SExpression => todo!(),
+            IonType::Struct => todo!(),
+        };
+
+        self.hasher.update(serialized_bytes);
+        self.hasher.finalize_reset()
```

Comment:
I think there needs to be a TODO for annotations.

Reply:
Added one, but I'm expecting the test suite to guide me here.